lib/eth/ens/resolver.rb

```diff
@@ -0,0 +1,78 @@
+# Copyright (c) 2016-2022 The Ruby-Eth Contributors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# -*- encoding : ascii-8bit -*-
+
+# Provides the {Eth} module.
+module Eth
+  # Provides ENS specific functionality
+  # ref: https://ens.domains
+  module Ens
+
+    # Utility class for resolving ENS names to Ethereum addresses
+    class Resolver
+
+      # The default address for ENS, which applies to most chains
+      DEFAULT_ADDRESS = '0x00000000000c2e074ec69a0dfb2997ba6c7d2e1e'.freeze
+
+      # Create an instance of the ENS Resolver
+      #
+      # @param ens_name [String] The ENS name, eg: fancy.eth
+      # @param ens_name [String] The ENS name, eg: fancy.eth
+      def initialize(client, address = DEFAULT_ADDRESS)
+        @client = client
+        @contract = Eth::Contract.from_abi(
```

Comment:
Few things here:

1) I would set only the contract filename here, so your object instantiation will not ever fail. Also set the ENS address as an instance variable. You may want to add a `valid?` predicate method, but not necessary.

```suggestion
        @address = address
        @contract_filename = "ens.json"
```

2. I would define a constant that points to the ABIs root directory.

```ruby
ABI_DIRECTORY = File.join(File.dirname(__FILE__), '../../../abis/ens.json'))
```

If you see this being reused, which I do, then probably add it to `Eth::Constants` instead.

3. Add a `def parsed_abi` method reads and parses the JSON file with error handling.

```ruby
def parsed_abi
  JSON.parse(File.read(File.join(ABI_DIRECTORY, @contract_address)
rescue StandardError => e
  # log something here
  puts "oh no, stuff blew up"
  @valid = false
end
```

4. Add a memoized `contract` method that instantiates the `Eth::Contract with the parsed ABI and address.

```ruby
def contract
  @contract ||= Eth::Contract.from_abi(name: 'ENS', address: address, abi: parsed_abi)
end
```

5. Use `attr_reader :address, :client, :parsed_abi` instead of accessing instance variables directly.

Reply:
I agree with the constant path for ABIs and error handling. #3 and #5 are conflicting, at least for the parsed_abi. Regarding attr accessors, these values are effectively private from the library user... but could be useful for debugging, I suppose.